Changes needed to pass the private CA while doing Reconciltion of Kafka enabled with SSL

rohitreddy1698 opened the pull request:
**Problem**
**TLS Certs for Kafka Not Initialised**
Currently the Milvus Operator does not have the provision to perform Validation Checks while connecting to an Externally Deployed Kafka which is SSL enabled and has certificates signed with a private CA.
When deployed the Milvus CR deployment fails stating that the **MsgStream is NotReady, and throws an SSL Certificate signed with unknown Authority Error** and the operator does not proceed to deploy the Milvus Components at all.

**Credentials being passed as plain strings**
Milvus components (librdkafka) require plain fields for Kafka auth/TLS:
`saslUsername`, `saslPassword`, `tlsCACert`, `tlsCert`, `tlsKey`, `tlsKeyPassword`.
The Operator today passes SecretRefs in `user.yaml`, and defaults can inject
Kerberos options, causing librdkafka failures during validation.

**What this PR does**
- Controller: resolves Kafka SecretRefs during reconcile into runtime-visible paths/strings.
- Runtime: enhances `scripts/run.sh` to materialize SASL/TLS from mounted Secret volumes
  (or env fallbacks) into the final `milvus.yaml` consumed by Milvus.
- Clears stray `kafka.properties` (e.g., accidental Kerberos defaults) to avoid validation errors.
- Keeps Secrets in Kubernetes Secrets; *does not* copy secret values into ConfigMaps/CRs.

**Why this approach**
- No Milvus core changes; all resolution happens in operator/runtime.
- Compatible with current Milvus config contract.
- Security: secrets remain in Secret volumes, only written in-pod to `milvus.yaml`.
- Enables reconciliation of Kafka signed with Private CA

**Backwards compatibility**
- If users already provide literal `saslUsername/saslPassword` or TLS file paths in the CR,
  those are respected (files win if Secret mount exists; mixed-mode supported).
- If only `caCertSecret` is provided, only `tlsCACert` is set (no accidental mTLS enable).

**How to use**
```yaml
spec:
  config:
    kafka:
      brokerList: ["broker:9094"]
      securityProtocol: SASL_SSL
      saslMechanisms: SCRAM-SHA-512
      saslUsernameSecret: { name: kafka-passwords, key: sasl-username }
      saslPasswordSecret: { name: kafka-passwords, key: sasl-password }
      ssl:
        enabled: true
        caCertSecret:  { name: kafka-ssl-secret, key: ca-cert }
        # (optional)
        certSecret:    { name: kafka-ssl-secret, key: client-cert }
        keySecret:     { name: kafka-ssl-secret, key: client-key }
        keyPasswordSecret: { name: kafka-passwords, key: tls-key-password }
```

Issue References : 
https://github.com/milvus-io/milvus/discussions/27977

Even though there is provision to configure the Kafka Certs in the Milvus CRD Definition file, the configured certificates are not being used and the TLS object is simply being initialised to an empty TLS object. Specifically here :
https://github.com/zilliztech/milvus-operator/blob/2a31cb6b2b0fa9054f2aa457d5b63b72d821fa70/pkg/external/kafka.go#L64

This PR contains the changes to address this issue, it has two changes :
- actually initialises the certs in`kafka.go` passed with the Milvus CRD.
- Controller: resolves Kafka SecretRefs during reconcile into runtime-visible paths/strings.
- Runtime: enhances `scripts/run.sh` to materialize SASL/TLS from mounted Secret volumes
  (or env fallbacks) into the final `milvus.yaml` consumed by Milvus.
- Clears stray `kafka.properties` (e.g., accidental Kerberos defaults) to avoid validation errors.
- Keeps Secrets in Kubernetes Secrets; *does not* copy secret values into ConfigMaps/CRs.

rohitreddy1698 commented:
Hello @haorenfsa 

Apologies for the delay in response and Thank you for your time for reviewing this and adding in your suggestions.

These were the suggestions you have given :
1)add an option to skip kafka checks
2)read the private CA from secrets

First one is simpler to me, and takes less time. And we may also need the same option for other dependencies - There should be an option to skip checks in Staging I understand but in production it might be necessary to ensure the health of dependencies. So I will maybe raise a Seperate PR providing an option to Skip Checks for all the dependencies.

2nd one is what you've be doing. But maybe the operator don't need to actually mount that secret, since it has the permission to read secret directly with kubernetes api access - I am currently taking in this approach.

Please let me know if you are okay with this, or have any other suggestions?

Thank you,
Rohit

rohitreddy1698 commented:
/assign @haorenfsa

rohitreddy1698 commented:
Hello @haorenfsa ,

Just checking in on this.
In case you got a chance to review this PR and any changes are necessary please do let me know.

Thank you !

haorenfsa commented:
Hi @rohitreddy1698 , one UT failed. And also the lint, seems due to using deprecated function of go lib

```
--- FAIL: TestGetKafkaDialer (0.00s)
    --- FAIL: TestGetKafkaDialer/securityProtocol=SASL_PLAINTEXT (0.00s)
        kafka_test.go:60: 
            	Error Trace:	/home/runner/work/milvus-operator/milvus-operator/pkg/external/kafka_test.go:60
            	Error:      	Received unexpected error:
            	            	SASL enabled ("") but username or password is empty (checked secrets and literals)
            	Test:       	TestGetKafkaDialer/securityProtocol=SASL_PLAINTEXT
```

rohitreddy1698 commented:
Hi @haorenfsa , @LoveEachDay

Can you please help review the code and let me know in case anything else is needed from my side for the PR to be merged.

Thank you !

haorenfsa commented on scripts/run.sh:
```diff
@@ -19,6 +19,85 @@ for (( i=0; i<$config_file_count; i++ )); do
+    fi
+  } | write_overlay
+fi
+
```
we should render it directly to the configmap in operator instead of doing merge while intializing

haorenfsa replied:
Ideally, a patch should be made to milvus to make it able to read the cert & secrets from certain file set in the config file.

haorenfsa replied:
And this script should only run when milvus configured to use kafka.

haorenfsa commented:
@rohitreddy1698 So 3 things there:
- [ ] The cert & key path should be rendered in controller's code instead of in the `run.sh`, the username&password part can be kept here.
- [ ] The script should only run when milvus uses kafka.
- [ ] It can be a temporary solution, but we should move the script to milvus' code afterwards.

Commit:
feat(kafka): materialize SASL/TLS from Kubernetes Secrets at runtime

- controllers: mount SecretRefs and render TLS file paths; add checksums for safe restarts
- scripts/run.sh: resolve SASL (and only when kafka enabled), avoid leaking secrets into ConfigMaps
- external/kafka.go(+tests): PKCS#8 support, SCRAM/PLAIN handling
- misc: lint fixes

Security: no secret material in ConfigMaps; read in-pod at runtime.
Signed-off-by: rohit-mothe <rmothe@apple.com>

rohitreddy1698 committed:
Merge branch 'main' into external-kafka-tls-private-ca

rohitreddy1698 committed:
Merge branch 'main' into external-kafka-tls-private-ca

rohitreddy1698 committed:
Merge branch 'main' into external-kafka-tls-private-ca

rohitreddy1698 committed:
Merge branch 'main' into external-kafka-tls-private-ca

rohitreddy1698 committed:
Merge branch 'main' into external-kafka-tls-private-ca

rohitreddy1698 commented:
Hello @haorenfsa 

Can you please help review the changes. 

Thank you !

rohitreddy1698 commented:
Hi @ezioguga

Can you please help review this and get this PR Closed ?

Thanks,
Rohit

rohitreddy1698 commented:
> @rohitreddy1698 So 3 things there:
> 
> * [x]  The cert & key path should be rendered in controller's code instead of in the `run.sh`, the username&password part can be kept here.
> * [x]  The script should only run when milus uses kafka.
> * [x]  It can be a temporary solution, but we should move the script to milvus' code afterwards.

Completed
